src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentPullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,195 @@
+        internal async override Task<DiagnosticReport[]> HandleRequestAsync(DocumentDiagnosticsParams request, ClientCapabilities clientCapabilities, string token, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (clientCapabilities is null)
+            {
+                throw new ArgumentNullException(nameof(clientCapabilities));
+            }
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            if (!documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpDoc))
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var referenceParams = new SerializableDocumentDiagnosticsParams()
+            {
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = csharpDoc.Uri
+                },
+                PreviousResultId = request.PreviousResultId,
+                PartialResultToken = token
+            };
+
+            if (!_lspProgressListener.TryListenForProgress(
+                token,
+                onProgressNotifyAsync: (value, ct) => ProcessDocumentDiagnosticsAsync(value, request.PartialResultToken, request.TextDocument.Uri, ct),
+                WaitForProgressNotificationTimeout,
+                cancellationToken,
+                out var onCompleted))
+            {
+                return null;
+            }
+
+            var result = await _requestInvoker.ReinvokeRequestOnServerAsync<SerializableDocumentDiagnosticsParams, DiagnosticReport[]>(
+                MSLSPMethods.DocumentPullDiagnosticName,
+                RazorLSPConstants.CSharpContentTypeName,
+                referenceParams,
+                cancellationToken).ConfigureAwait(false);
+
+            // We must not return till we have received the progress notifications
+            // and reported the results via the PartialResultToken
+            await onCompleted.ConfigureAwait(false);
+
+            return null;
+        }
+
+        private async Task ProcessDocumentDiagnosticsAsync(
+            JToken value,
+            IProgress<DiagnosticReport[]> progress,
+            Uri razorDocumentUri,
+            CancellationToken cancellationToken)
+        {
+            var result = value.ToObject<DiagnosticReport[]>();
+
+            if (result == null || result.Length == 0)
+            {
+                return;
+            }
+
+            var remappedResults = await RemapDocumentDiagnosticsAsync(result, razorDocumentUri, cancellationToken).ConfigureAwait(false);
+
+            progress.Report(remappedResults);
+        }
+
+        private async Task<DiagnosticReport[]> RemapDocumentDiagnosticsAsync(
```

Comment:
Hmm, we should add a "RemapCSharpDiagnostics" endpoint to our language server so we can do this all in one go everywhere the language server is running. I know it isn't 100% straightforward because our language server doesn't have a concept of the path of the generated C# file so it'd have to be some sort of contract that talks in only diagnostic ids/names and locations/ranges. It'd enable us to do it for VSCode but also it'd be more performant instead of remappping a bajillion diagnostics 😄 

Similarily we'll also add a FilterHTMLDiagnostics" when we add that support for embedded lang diagnostics.

Reply:
Would you imagine this to be part of this PR, or addressed separately?

Reply:
Actually...

We may just be able to leverage the existing batch mapping logic to minimize the number of calls as we do with the highlight handler.

https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentHighlightHandler.cs#L107-L136

Reply:
> Would you imagine this to be part of this PR, or addressed separately?

However you feel is appropriate. One other motivation to having it on the language server is to allow removal of certain diagnostics. Such as "Unnecessary using xyz" for _Imports.razor files. There's a few other warnings too that we'll need to remove.

Reply:
Okay, so I've moved over to using batch diagnostics via https://github.com/dotnet/aspnetcore-tooling/pull/2709/commits/3b0ee6a4905cd01e2f42a1bfabbafdbdf2f40486. Regarding the larger move to the language server and the creation of the new endpoint, I created https://github.com/dotnet/aspnetcore/issues/27563 to track the issue. My goal is to get this merged ASAP so we can all start trying it out and identify outstanding issues.